### PR TITLE
ServerList implementation for Eureka2

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public List<DiscoveryEnabledServer> getInitialListOfServers() {
+        return loadListOfServers();
+    }
+
+    @Override
+    public List<DiscoveryEnabledServer> getUpdatedListOfServers() {
+        return loadListOfServers();
+    }
+
+    private List<DiscoveryEnabledServer> loadListOfServers() {
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java]: Does this mean flipping from eureka 2 to eureka 1 & vice-versa requires reboot of an application?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    public List<DiscoveryEnabledServer> getUpdatedListOfServers() {
+        return obtainServersViaEureka2();
+    }
+
+    /**
+     * TODO prioritizeVipAddressBasedServers not supported yet; all vips are subscribed to eagerly
+     */
+    private List<DiscoveryEnabledServer> obtainServersViaEureka2() {
+        logger.info("Resolving {} from Eureka2...", this.vipAddresses);
+        List<InstanceInfo> instanceInfos = getServerListReader().getLatestServerListOrWait();
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Is there a timeout for this wait or it waits forever?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java (hunk 1 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java]: There is no need for reboot. For each request a status flag is checked, and the request is dispatched accordingly.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 2 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Yes, and it is hard coded to 30sec now. If you feel it is better to make it configurable, I will add one more property to cover that.

Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: I think from eureka point of view it should be configurable (passed as an argument to this method) , whether we make it configurable from ribbon or not, it is up to you.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java (hunk 1 above)
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java]: Cool :+1: 

Does ribbon overwrites the new list of servers? How would `getInitialListOfServers()` from eureka1 and `getUpdatedListOfServers()` from eureka2 work?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    private Eureka2Clients() {
+    }
+
+    /**
+     * In case both Eureka1 and Eureka2 clients are available, indicate which one is a preferred one,
+     * This flag is read by {@link CompositeEurekaEnabledNIWSServerList} to decide to which client discard
+     * the request.
+     */
+    public static boolean isPreferEureka2() {
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Nitpicking: Is this "prefer" or "use"?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    public static void setPreferEureka2(boolean peferEureka2) {
+        Eureka2Clients.peferEureka2 = peferEureka2;
+    }
+
+    public static EurekaInterestClient getInterestClient() {
+        return interestClient;
+    }
+
+    public static void setInterestClient(EurekaInterestClient interestClient) {
+        Eureka2Clients.interestClient = interestClient;
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Is there a need to shutdown the existing client, if present?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+                ServerResolver readClusterResolver = ServerResolvers
+                        .fromEureka(writeClusterResolver)
+                        .forInterest(Interests.forVips(readClusterVip));
+
+                interestClient = Eurekas.newInterestClientBuilder().withServerResolver(readClusterResolver).build();
+            }
+
+            String[] vipAddresses = this.vipAddresses.split(",");
+            serverListReaderRef.compareAndSet(null, new ServerListReader(interestClient, vipAddresses, isSecure));
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Shutdown the new reader if CAS failed?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    private String datacenter;
+
+    private int overridePort = DefaultClientConfigImpl.DEFAULT_PORT;
+    private boolean shouldUseOverridePort;
+    private boolean shouldUseIpAddr;
+
+    private AtomicReference<ServerListReader> serverListReaderRef = new AtomicReference<ServerListReader>();
+
+    @Override
+    public void initWithNiwsConfig(IClientConfig clientConfig) {
```
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Are any of these configs dynamically changeable? Looks like we load it once.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java (hunk 1 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/CompositeEurekaEnabledNIWSServerList.java]: I guess it is up to load balancer implementation, but what is currently there (BaseLoadBalancer) swaps old list with a new one. In any case, the data we get from Eureka1 and Eureka2 should be eventually the same.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 2 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: On a second thought I will make it configurable via property. Some tuning should be possible.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java (hunk 3 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: "use" is better word here. I have renamed it.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java (hunk 4 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: It depends how we define the contract. Now Eureka2Clients uses injected EurekaInterestClient, but does not own it. In DI environment EurekaInterestClientProvider is managing the full lifecycle (create and destroy). I think it is better as single client can be used in different places.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 5 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Good catch.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 6 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: This is how we do it in DiscoveryEnabledNIWSServerList. Client configuration change should be handled by creating a new ribbon client, for example If there is a vip change, RestClientManager will create a new RestClient.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 2 above)
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: I think it is not at all intuitive if the timeout is configured via property inside `ServerListReader` and more so it is global value then for all awaits.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 6 above)
Comment (NiteshKant) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: ok

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 2 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: `ServerListReader` has a constructor argument for this timeout, and in Ribbon I have introduced a configuration property.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 6 above)
Comment (qiangdavidliu) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: This method is the same code as what's in DiscoveryEnabledNIWSServerList. Can we consolidate commonalities in both into a base abstract class?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        logger.info("Resolving {} from Eureka2...", this.vipAddresses);
+        List<InstanceInfo> instanceInfos = getServerListReader().getLatestServerListOrWait();
+        if (instanceInfos == null) {
+            logger.warn("Server resolve for vip={},secure={} timed out", this.vipAddresses, isSecure);
+            return Collections.emptyList();
+        }
+        return toServerList(instanceInfos);
+    }
+
+    private ServerListReader getServerListReader() {
```
Comment (qiangdavidliu) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Do we need to ensure thread safety for this method?

Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: This method is thread safe, which comes from thread safety of ServerListReader.

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+     * In case both Eureka1 and Eureka2 clients are available, indicate which one should be used.
+     * This flag is read by {@link CompositeEurekaEnabledNIWSServerList} to decide to which client discard
+     * the request.
+     */
+    public static boolean isUseEureka2() {
+        return useEureka2;
+    }
+
+    public static boolean setUseEureka2(boolean useEureka2) {
+        boolean lastValue = Eureka2Clients.useEureka2;
```
Comment (elandau) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Why not just use AtomicBoolean.getAndSet()?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        Eureka2Clients.useEureka2 = useEureka2;
+        return lastValue;
+    }
+
+    public static EurekaInterestClient getInterestClient() {
+        return interestClient;
+    }
+
+    public static EurekaInterestClient setInterestClient(EurekaInterestClient interestClient) {
```
Comment (elandau) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Why not use AtomicReference.getAndSet()?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+        ServerListReader current = serverListReaderRef.get();
+        if (current != null && current.isClosed()) {
+            serverListReaderRef.compareAndSet(current, null);
```
Comment (elandau) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Return value of compareAndSet never used?  Should it be?

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java (hunk 8 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Changed

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java (hunk 9 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2Clients.java]: Changed

File: ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java (hunk 10 above)
Comment (tbak) [inline ribbon-eureka/src/main/java/com/netflix/niws/loadbalancer/Eureka2EnabledNIWSServerList.java]: Yes, it is correct. We want to set it to null, only if current object is there. As this method may be run concurrently, another thread may initialize new serverListReader at the same time.

**Comment by tbak** [comment]
The build failure was caused by DNS resolution problem in one test.

**Pull request merged**